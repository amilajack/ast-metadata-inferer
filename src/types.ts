import { CompatStatement } from "mdn-browser-compat-data/types";

export type ApiMetadata<T = Language.JS> = {
  id: string;
  name: string;
  language: T;
  kind: APIKind;
  protoChain: Array<string>;
  protoChainId: string;
  compat: CompatStatement;
};

export type CssApiMetadata = ApiMetadata<Language.CSS>;

export type JsApiMetadata = ApiMetadata<Language.JS>;

export enum Language {
  JS = "js-api",
  CSS = "css-api",
}

export enum APIKind {
  Web = "web",
  ES = "es",
}
