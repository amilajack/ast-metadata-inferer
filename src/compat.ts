import fs from "fs";
import path from "path";
import AstMetadataInferer from ".";
import MdnComaptDataProvider from "./providers/MdnCompatDataProvider";
import { RecordType } from "./types";

export default async function Compat(): Promise<RecordType> {
  const records = await AstMetadataInferer();
  // Add all the corresponding compat data for each inferred ast node
  const compatDataMap = new Map(
    MdnComaptDataProvider().map((e) => [e.protoChainId, e])
  );
  const apisWithCompatRecords = records.filter((api) =>
    compatDataMap.has(api.protoChainId)
  );

  const compatRecordsFile = path.join(__dirname, "..", "compat.json");
  await fs.promises.writeFile(
    compatRecordsFile,
    JSON.stringify(apisWithCompatRecords)
  );

  return apisWithCompatRecords;
}

if (require.main === module) {
  Compat().catch((e) => {
    throw new Error(e);
  });
}
