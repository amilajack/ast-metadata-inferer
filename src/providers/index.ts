import mdnCompatData from "./mdn";
// @TODO: Needs to return compat records
// import MsApiCatalogProvider from "./MsApiCatalogProvider";
import type { ApiMetadata } from "../types";

export default async function Providers(): Promise<Array<ApiMetadata>> {
  const [mdnRecords] = await Promise.all([mdnCompatData()]);
  const map: Map<string, ApiMetadata> = new Map<string, ApiMetadata>(
    mdnRecords.map((record) => [record.protoChainId, record])
  );

  return Array.from(map.values()).filter(
    (record) =>
      !record.protoChain.includes("RegExp") &&
      !record.protoChainId.includes("@@")
  );
}
