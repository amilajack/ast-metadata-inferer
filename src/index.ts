import fs from "fs";
import path from "path";
import providers from "./providers";
import astNodeTypesTester from "./helpers/ast-node-types-tester";
import { ApiMetadata } from "./types";

const API_BLACKLIST = ["close", "confirm", "print"];

export default async function astMetadataInferer(): Promise<ApiMetadata[]> {
  // @HACK: Temporarily ignoring the last 1K records because they
  //        cause issues for some unknown reason. They prevent
  //        AstMetadataInferer from returning
  const providerResults = await providers();
  const records = providerResults.filter(
    (metadata) => !API_BLACKLIST.includes(metadata.name)
  );
  const file = path.join(__dirname, "..", "metadata.json");

  if (fs.existsSync(file)) {
    await fs.promises.unlink(file);
  }

  const promises = [];
  const parallelisim = 4;
  const eachRecordsSize = Math.floor(records.length / parallelisim);

  for (let i = 0; i < parallelisim; i += 1) {
    const recordsSliceEnd =
      i === parallelisim ? records.length + 1 : (i + 1) * eachRecordsSize;
    const recordsSlice = records.slice(i * eachRecordsSize, recordsSliceEnd);
    promises.push(astNodeTypesTester(recordsSlice));
  }

  const recordsWithMetadata = await Promise.all(promises).then((res) =>
    res.reduce((p, c) => p.concat(c), [])
  );

  await fs.promises.writeFile(file, JSON.stringify(recordsWithMetadata));

  return recordsWithMetadata;
}
