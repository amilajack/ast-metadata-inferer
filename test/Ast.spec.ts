/* eslint global-require: off, no-console: off */
import fs from "fs";
import path from "path";
import astMetadataInferer from "../src";
import { ApiMetadata } from "../src/types";

jest.setTimeout(60000);

describe("AstMetadataInferer", () => {
  it("should return basic output", async () => {
    const [record, ...records] = await astMetadataInferer();
    expect(record).toHaveProperty("apiType");
    expect(record).toHaveProperty("type");
    expect(record).toHaveProperty("protoChain");
    expect(record).toHaveProperty("protoChainId");
    expect(record).toHaveProperty("astNodeTypes");
    expect(record).toHaveProperty("isStatic");
    expect(record).toMatchSnapshot();

    records.forEach((_record) => {
      expect(_record).toHaveProperty("apiType");
      expect(_record).toHaveProperty("type");
      expect(_record).toHaveProperty("protoChain");
      expect(_record).toHaveProperty("protoChainId");
      expect(_record).toHaveProperty("astNodeTypes");
      expect(_record).toHaveProperty("isStatic");
    });
  });

  it("should write to metadata.json correctly", async () => {
    const filepath = path.join(__dirname, "..", "metadata.json");
    const file = await fs.promises.readFile(filepath);

    expect(JSON.parse(file.toString())[0]).toMatchSnapshot();
    const recordsCount = JSON.parse(file.toString()).length;
    expect(recordsCount).toBeGreaterThanOrEqual(4000);

    console.log(`${recordsCount} records in metadata.json`);

    const astMetadata = await astMetadataInferer();
    expect(astMetadata).toHaveLength(recordsCount);
  });

  it("should expose metadata.json in parsable format", async () => {
    const astMetadata = await astMetadataInferer();
    const querySelectorRecord = astMetadata.find(
      (apiMetadata: ApiMetadata) =>
        apiMetadata.protoChainId === "document.querySelector"
    );
    expect(querySelectorRecord).toMatchSnapshot();
  });
});
