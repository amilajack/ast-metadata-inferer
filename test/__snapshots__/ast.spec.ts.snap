// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstMetadataInferer should expose metadata.json in parsable format 1`] = `
Object {
  "astNodeTypes": Array [
    "MemberExpression",
  ],
  "compat": Object {
    "mdn_url": "https://developer.mozilla.org/docs/Web/API/Document/querySelector",
    "status": Object {
      "deprecated": false,
      "experimental": false,
      "standard_track": true,
    },
    "support": Object {
      "chrome": Object {
        "version_added": "1",
      },
      "chrome_android": Object {
        "version_added": "18",
      },
      "edge": Object {
        "version_added": "12",
      },
      "firefox": Object {
        "version_added": "3.5",
      },
      "firefox_android": Object {
        "version_added": "4",
      },
      "ie": Object {
        "version_added": "8",
      },
      "opera": Object {
        "version_added": "10",
      },
      "opera_android": Object {
        "version_added": "10.1",
      },
      "safari": Object {
        "version_added": "3.2",
      },
      "safari_ios": Object {
        "version_added": "3",
      },
      "samsunginternet_android": Object {
        "version_added": "1.0",
      },
      "webview_android": Object {
        "version_added": "1",
      },
    },
  },
  "id": "document.querySelector",
  "isStatic": true,
  "kind": "web",
  "language": "js-api",
  "name": "querySelector",
  "protoChain": Array [
    "document",
    "querySelector",
  ],
  "protoChainId": "document.querySelector",
}
`;

exports[`AstMetadataInferer should return basic output 1`] = `
Object {
  "astNodeTypes": Array [
    "NewExpression",
  ],
  "compat": Object {
    "mdn_url": "https://developer.mozilla.org/docs/Web/API/AbortController",
    "status": Object {
      "deprecated": false,
      "experimental": true,
      "standard_track": true,
    },
    "support": Object {
      "chrome": Object {
        "version_added": "66",
      },
      "chrome_android": Object {
        "version_added": "66",
      },
      "edge": Object {
        "version_added": "16",
      },
      "firefox": Object {
        "version_added": "57",
      },
      "firefox_android": Object {
        "version_added": "57",
      },
      "ie": Object {
        "version_added": false,
      },
      "opera": Object {
        "version_added": "53",
      },
      "opera_android": Object {
        "version_added": "47",
      },
      "safari": Array [
        Object {
          "version_added": "12.1",
        },
        Object {
          "notes": "Even though <code>window.AbortController</code> is defined, it doesn't really abort <code>fetch</code> requests. See <a href='https://webkit.org/b/174980'>bug 174980</a>.",
          "partial_implementation": true,
          "version_added": "11.1",
        },
      ],
      "safari_ios": Array [
        Object {
          "version_added": "12.2",
        },
        Object {
          "notes": "Even though <code>window.AbortController</code> is defined, it doesn't really abort <code>fetch</code> requests. See <a href='https://webkit.org/b/174980'>bug 174980</a>.",
          "partial_implementation": true,
          "version_added": "11.3",
        },
      ],
      "samsunginternet_android": Object {
        "version_added": "9.0",
      },
      "webview_android": Object {
        "version_added": "66",
      },
    },
  },
  "id": "AbortController",
  "isStatic": true,
  "kind": "web",
  "language": "js-api",
  "name": "AbortController",
  "protoChain": Array [
    "AbortController",
  ],
  "protoChainId": "AbortController",
}
`;

exports[`AstMetadataInferer should write to metadata.json correctly 1`] = `
Object {
  "astNodeTypes": Array [
    "NewExpression",
  ],
  "compat": Object {
    "mdn_url": "https://developer.mozilla.org/docs/Web/API/AbortController",
    "status": Object {
      "deprecated": false,
      "experimental": true,
      "standard_track": true,
    },
    "support": Object {
      "chrome": Object {
        "version_added": "66",
      },
      "chrome_android": Object {
        "version_added": "66",
      },
      "edge": Object {
        "version_added": "16",
      },
      "firefox": Object {
        "version_added": "57",
      },
      "firefox_android": Object {
        "version_added": "57",
      },
      "ie": Object {
        "version_added": false,
      },
      "opera": Object {
        "version_added": "53",
      },
      "opera_android": Object {
        "version_added": "47",
      },
      "safari": Array [
        Object {
          "version_added": "12.1",
        },
        Object {
          "notes": "Even though <code>window.AbortController</code> is defined, it doesn't really abort <code>fetch</code> requests. See <a href='https://webkit.org/b/174980'>bug 174980</a>.",
          "partial_implementation": true,
          "version_added": "11.1",
        },
      ],
      "safari_ios": Array [
        Object {
          "version_added": "12.2",
        },
        Object {
          "notes": "Even though <code>window.AbortController</code> is defined, it doesn't really abort <code>fetch</code> requests. See <a href='https://webkit.org/b/174980'>bug 174980</a>.",
          "partial_implementation": true,
          "version_added": "11.3",
        },
      ],
      "samsunginternet_android": Object {
        "version_added": "9.0",
      },
      "webview_android": Object {
        "version_added": "66",
      },
    },
  },
  "id": "AbortController",
  "isStatic": true,
  "kind": "web",
  "language": "js-api",
  "name": "AbortController",
  "protoChain": Array [
    "AbortController",
  ],
  "protoChainId": "AbortController",
}
`;
